drivers/soundwire/bus.c

```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
 	return ret;
 }
 
+static int sdw_slave_set_frequency(struct sdw_slave *slave)
+{
+	u32 mclk_freq = slave->bus->prop.mclk_freq;
+	u32 curr_freq = slave->bus->params.curr_dr_freq >> 1;
+	u8 base_offset = 0;
+	u8 base = 0;
+	u8 scale_index = 0;
+	u32 scale;
+	int ret;
+	int i;
+
+	/*
+	 * frequency base and scale registers are required for SDCA
+	 * devices. They may also be used for 1.2+/non-SDCA devices,
+	 * but we will need a DisCo property to cover this case
+	 */
+	if (!slave->id.class_id)
+		return 0;
+
+	if (!mclk_freq) {
+		dev_err(&slave->dev,
+			"no bus MCLK, cannot set SDW_SCP_BUS_CLOCK_BASE\n");
+		return -EINVAL;
+	}
+
+	/* map base frequency using Table 89 of SoundWire 1.2 spec */
+	if (!(38400000 % mclk_freq)) {
+		base = SDW_SCP_BASE_CLOCK_19200000_HZ;
+		base_offset = 1;
```

Comment:
Hmm, this is slightly hard to follow. If we assume sane mclk_freq value range, then the algorithm seems to work though, so ok. E.g. mclk_freq of 0.6Mhz would lead to base_offset=1 and base=24Mhz, but that would fail in trying to find a suitable scaling factor. So looks ok, although a bit hard to review.

Reply:
mclk is a property in the BIOS, it's currently:

"intel-sdw-ip-clock", 
0x016E3600 -> 24
0x0249F000 -> 38,4

The 1 is because MIPI defines 19.2 so we have a factor to correct.
Will add some more comments here.